curriculum/challenges/english/25-front-end-development/workshop-fcc-authors-page/641da615af82bf454215a992.md

```diff
@@ -0,0 +1,164 @@
+---
+id: 641da615af82bf454215a992
+title: Step 10
+challengeType: 0
+dashedName: step-10
+---
+
+# --description--
+
+Pass an empty callback function to the `.forEach()` method. For the first parameter of the callback, destructure the `author`, `image`, `url`, and `bio` items. 
+
+For the second parameter, pass in `index`. This will represent the position of each author, and will be useful for pagination later.
+
+# --hints--
+
+You should destructure `author`, `image`, `url`, and `bio` as the first parameter of your callback function. Don't forget to put them inside curly braces.
+
+```js
+assert.match(code, /authors\.forEach\(\s*\(\s*\{\s*(?:author\s*,\s*image\s*,\s*url\s*,\s*bio)|(?:image\s*,\s*author\s*,\s*url\s*,\s*bio)|(?:url\s*,\s*author\s*,\s*image\s*,\s*bio)|(?:author\s*,\s*url\s*,\s*image\s*,\s*bio)|(?:image\s*,\s*url\s*,\s*author\s*,\s*bio)|(?:url\s*,\s*image\s*,\s*author\s*,\s*bio)|(?:url\s*,\s*image\s*,\s*bio\s*,\s*author)|(?:image\s*,\s*url\s*,\s*bio\s*,\s*author)|(?:bio\s*,\s*url\s*,\s*image\s*,\s*author)|(?:url\s*,\s*bio\s*,\s*image\s*,\s*author)|(?:image\s*,\s*bio\s*,\s*url\s*,\s*author)|(?:bio\s*,\s*image\s*,\s*url\s*,\s*author)|(?:bio\s*,\s*author\s*,\s*url\s*,\s*image)|(?:author\s*,\s*bio\s*,\s*url\s*,\s*image)|(?:url\s*,\s*bio\s*,\s*author\s*,\s*image)|(?:bio\s*,\s*url\s*,\s*author\s*,\s*image)|(?:author\s*,\s*url\s*,\s*bio\s*,\s*image)|(?:url\s*,\s*author\s*,\s*bio\s*,\s*image)|(?:image\s*,\s*author\s*,\s*bio\s*,\s*url)|(?:author\s*,\s*image\s*,\s*bio\s*,\s*url)|(?:bio\s*,\s*image\s*,\s*author\s*,\s*url)|(?:image\s*,\s*bio\s*,\s*author\s*,\s*url)|(?:author\s*,\s*bio\s*,\s*image\s*,\s*url)|(?:bio\s*,\s*author\s*,\s*image\s*,\s*url)\s*\}/)
```

Comment:
Just looking at it, it's an absolute monster of a check. I swear we should probably have a helper to make sure the elements are present but not necessarily in the same order as we expect.

Reply:
Yeah, it's a monster. And regex tests should be replaced when possible. I think we agreed on doing it in 2025 for a matter of time.

Reply:
I have a silly question. How much of my initial function need to be modified to take care of a forEach statement instead of a function? https://github.com/freeCodeCamp/curriculum-helpers/blob/main/lib/__tests__/javascript-helper.test.ts

Reply:
I'm not sure, but considering that the forEach can be chained to something of variable structure it could be a mess.

>  https://github.com/freeCodeCamp/curriculum-helpers/blob/main/lib/__tests__/javascript-helper.test.ts

Do we use these anywhere?

Reply:
Could we try to capture parameters of the callback, then sort them and compare with single array?

Reply:
I actually don't think we will need that helper in this case, because these steps are broken down to much. I understand why that was in the current context. But for the new curriculum, it won't need to be broken down to this extent. 

In the future refactor PR's, we will probably just have one step to create the initial `displayAuthors` function where the author container is updated with this user card div

```js
 <div id="${index}" class="user-card">
      <h2 class="author-name">${author}</h2>
 </div>
``` 

Campers will have a much more solid foundation before they reach this project, so a lot of these steps will be more opened ended similar to labs, and it will be 10X easier to test because we will testing the final result instead of using all of this regex.

But for now, we can just merge this project as is because it is stable enough to go in.
Then we can worry about refactors in future PR's.

Even the most hardcode campers won't reach this project until maybe Feburary. (if that). Most won't reach it until, halfway through the year because of the way the new curriculum is setup.